### DISCUSSION TITLE
Installation on RPi4 ?

### DISCUSSION THREAD

**Discussion by hoschiking** [DennisOSRM/hms-mqtt-publisher#101] [Q&A]
It is possible to install this tool on a Raspberry Pi 4 with Raspberry Pi OS?
If not, are you going to offer this possibility?
thx

**Comment by dentech82** [comment]
I've installed the script on a Raspberry 3 B+ with Raspberry OS, the installation on a Raspberry 4 should also work.

**Reply by hoschiking** [reply]
thx for the answer.
I installed and compiled it and it works, but only one time.
when I close the CLI, the tool stops.
how can I start the tool manually?
what is the right way to automatic start it after reboot?
sorry for the questions, I'm a linux beginner, software is my handycap.
thank you for the support

**Reply by dentech82** [reply]
See here: https://github.com/DennisOSRM/hms-mqtt-publisher/discussions/95#discussion-6214683

**Comment by topcaser** [comment]
Hi, how can I compile that on a Raspi 5 with Raspian OS?

**Reply by dentech82** [reply]
There is nothing to compile. Create the script with nano, that's all.

**Comment by DennisOSRM** [comment]
Compiling from source is easy, just install the Rust toolchain [link](https://rustup.rs/). Then check out the code and run `cargo r` to compile and start.

**Comment by hoschiking** [comment]
For first I tested it on my test RPi5, it works fine.
Now I wanted to install it on my RPi4, but it is not possible.
I got the failer:

error: failed to parse manifest at `/home/pi/hms-mqtt-publisher/Cargo.toml`

Caused by:
  failed to parse the `edition` key

Caused by:
  this version of Cargo is older than the `2021` edition, and only supports `2015` and `2018` editions.

The System is Raspberry OS Bullseye

How can I fix it?

**Reply by DennisOSRM** [reply]
Sounds like a rather old version of the rust toolchain. Try `rustup update` to upgrade to the latest one.

**Reply by hoschiking** [reply]
I installed it today and can´t install any newer version over "apt-get install"
I think, all another Rpi4 user should have the same problem........

**Reply by DennisOSRM** [reply]
Remove the one coming with the OS since it is apparently pretty old. Instead install the toolchain via rustup: [link](https://rustup.rs/)

**Reply by hoschiking** [reply]
I install it, but now i got a new error *confused*
cargo r
error: command failed: 'cargo': No such file or directory (os error 2)

**Reply by DennisOSRM** [reply]
try logging out and back in

**Reply by hoschiking** [reply]
I did it do, I rebooted the Rpi.
I checked the path, the path ~/.cargo/bin exists.....
It is necessery to to define the path in the config? If yes, where?
Sorry for the questions

**Reply by DennisOSRM** [reply]
I have this line at the end of my `.profile` to set environment variables:

```
. "$HOME/.cargo/env"
```

**Reply by hoschiking** [reply]
this line exists in .profile
I removed all (rust etc...) and installed it one more time
this error is present.... :-(
